mentoring/components/answer.py

```diff
@@ -0,0 +1,268 @@
+# General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program in a file in the toplevel directory called
+# "AGPLv3".  If not, see <http://www.gnu.org/licenses/>.
+#
+
+# Imports ###########################################################
+
+import logging
+from lazy import lazy
+
+from mentoring.models import Answer
+
+from xblock.core import XBlock
+from xblock.fields import Scope, Float, Integer, String
+from xblock.fragment import Fragment
+from xblock.validation import ValidationMessage
+from xblockutils.resources import ResourceLoader
+from xblockutils.studio_editable import StudioEditableXBlockMixin
+from .step import StepMixin
+import uuid
+
+
+# Globals ###########################################################
+
+log = logging.getLogger(__name__)
+loader = ResourceLoader(__name__)
+
+# Classes ###########################################################
+
+
+class AnswerMixin(object):
+    """
+    Mixin to give an XBlock the ability to read/write data to the Answers DB table.
+    """
+    def _get_course_id(self):
+        """ Get a course ID if available """
+        return getattr(self.runtime, 'course_id', 'all')
+
+    def _get_student_id(self):
+        """ Get student anonymous ID or normal ID """
+        try:
+            return self.runtime.anonymous_student_id
+        except AttributeError:
+            return self.scope_ids.user_id
+
+    def get_model_object(self, name=None):
+        """
+        Fetches the Answer model object for the answer named `name`
+        """
+        # By default, get the model object for the current answer's name
+        if not name:
+            name = self.name
+        # Consistency check - we should have a name by now
+        if not name:
+            raise ValueError('AnswerBlock.name field need to be set to a non-null/empty value')
+
+        student_id = self._get_student_id()
+        course_id = self._get_course_id()
+
+        answer_data, _ = Answer.objects.get_or_create(
+            student_id=student_id,
+            course_id=course_id,
+            name=name,
+        )
+        return answer_data
+
+    def validate_field_data(self, validation, data):
+        """
+        Validate this block's field data.
+        """
+        super(AnswerMixin, self).validate_field_data(validation, data)
+
+        def add_error(msg):
+            validation.add(ValidationMessage(ValidationMessage.ERROR, msg))
+
+        if not data.name:
+            add_error(u"A Question ID is required.")
+
+
+class AnswerBlock(AnswerMixin, StepMixin, StudioEditableXBlockMixin, XBlock):
+    """
+    A field where the student enters an answer
+
+    Must be included as a child of a mentoring block. Answers are persisted as django model instances
+    to make them searchable and referenceable across xblocks.
+    """
+    name = String(
+        display_name="Question ID (name)",
```

Comment:
It is sort of being auto-generated - see `@classmethod def get_template` below in this class. Basically in Studio, it will generate a new ID as part of the "Add a New Block" button process. In Workbench/XML you must manually specify an ID.